blog/_src/posts/2018-12-02-java-transient.md

```diff
@@ -0,0 +1,361 @@
+    Title: Java and Migratory Typing
+    Date: 2018-12-02T14:41:53
+    Tags: migratory typing, gradual typing, java, by Ben Greenman
+
+The _transient_ approach to migratory typing (circa [2014](http://homes.sice.indiana.edu/mvitouse/papers/dls14.pdf))
+ is similar to type erasure in Java (circa [2004](https://docs.oracle.com/javase/1.5.0/docs/relnotes/features.html)).
+
+<!-- more -->
+
+## Migratory typing
+
+The goal of _migratory typing_ is to enrich the type system of a language
+ without breaking backwards compatibility.
+Ideally, code that uses the enriched types:
+
+- (G1) benefits from new ahead-of-time checks,
+- (G2) benefits from stronger run-time guarantees, and
+- (G3) may interact with all kinds of existing code.
+
+There are tradeoffs involved in the implementation of a migratory typing
+ system, however, and (as we will see) different implementations may focus on
+ different goals that the three above.
+
+A typical migratory typing system adds a static type checker to a dynamically
+ typed language ([examples](/blog/2018/10/06/a-spectrum-of-type-soundness-and-performance/index.html)),
+ but one could also extend the type system of a statically-typed language
+ (e.g., by [adding dependent types](https://hal.inria.fr/hal-01629909v2)).
+In this sense, Java 1.5.0 is a migratory typing system for pre-generics Java.
+The addition of generic types enabled new ahead-of-time checks and maintained backwards
+ compatibility with existing Java code.
+
+Java's implementation of migratory typing has some interesting things in common
+ with the _transient_ implementation strategy recently proposed by
+ Michael Vitousek and collaborators.
+The goal of this post is to demonstrate the connections.
+
+
+## Erasure migratory typing
+
+Before we compare Java 1.5.0 to transient, let's review a strategy that
+ pre-dates and informs them both: the _erasure_ approach to migratory typing.
+
+[TypeScript](https://www.typescriptlang.org/) is a great example of the erasure approach.
+TypeScript is a migratory typing system for JavaScript.
+A TypeScript module gets validated by an ahead-of-time type checker and
+ compiles to JavaScript.
+After compilation, any JavaScript program may import bindings
+ from the generated code.
+Conversely, a TypeScript module may import bindings from a JavaScript module
+ by declaring a static type for each binding ([click for lots of examples](http://definitelytyped.org/)).
```

Comment:
it might help to give a clearer idea of what the link is, like "Joe Schmo of Definitely Typed curates lots of excellent examples of what this looks like"